### PR TITLE
KNOX-2402 - Adding Gateway performance testing

### PR DESCRIPTION
**Pull request by smolnar82**
## What changes were proposed in this pull request?

All is described [here](https://cwiki.apache.org/confluence/display/KNOX/KNOX-2402+-+Develop+a+performance-test+framework+to+test+token+state+service).

## How was this patch tested?

Manual testing: configured a Knox Gateway and executed the performance test tool multiple times. I checked the generated test results as well as the log files. See the above-referenced link for more information.
I even revealed some performance test issues in `AliasBasedTokenStateService` (see [KNOX-2408](https://issues.apache.org/jira/browse/KNOX-2408))

### CODE REVIEW

**Comment by smolnar82** [comment]
I still need to produce documentation about the tool (and update the PR's description once it's done). I just wanted you guys to review the code as soon as possible.

**Comment by smolnar82** [comment]
Thanks, @risdenk for your comment! It totally makes sense; there is no reason to overcomplicate it. I added a new profile to run performance tests using Maven instead of having a separate release artifact.

File: gateway-performance-test/pom.xml [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+					                    <systemProperty>
+					                        <key>perf.test.report.generation.target.folder</key>
+					                        <value>${basedir}/target/testResults</value>
+					                    </systemProperty>
+					                    <systemProperty>
+					                        <key>log4j.configuration</key>
+					                        <value>file:${basedir}/src/test/resources/performanceTest-log4j.properties</value>
+					                    </systemProperty>
+					                </systemProperties>
+					            </configuration>
```
[the remark this answers is not in the document]
Comment (smolnar82) [reply gateway-performance-test/pom.xml]: Ohh...copy-pasted that section from the original Maven page which comes with tabs. Fixed now.

File: gateway-performance-test/src/main/java/org/apache/knox/gateway/performance/test/PerformanceTestRunner.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+public class PerformanceTestRunner {
+
+  public static void main(String[] args) throws Exception {
+    final PerformanceTestConfiguration configuration = new PerformanceTestConfiguration(args[0]);
+    final ResponseTimeCache responseTimeCache = new ResponseTimeCache();
+    final GatewayMetricsReporter metricsReporter = new GatewayMetricsReporter(configuration, responseTimeCache);
+    metricsReporter.start();
+    final List<PerformanceTestLifeCyleListener> lifeCyleListeners = Arrays.asList(metricsReporter);
+    for (PerformanceTestUseCase useCase : PerformanceTestUseCase.values()) {
```
Comment (pzampino) [inline gateway-performance-test/src/main/java/org/apache/knox/gateway/performance/test/PerformanceTestRunner.java:33]: Could use ServiceLoader instead of a hard-coded enum to find PerformanceTestUseCase implementations.

File: gateway-performance-test/src/main/java/org/apache/knox/gateway/performance/test/knoxtoken/KnoxTokenCache.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+class KnoxTokenCache {
+  private static final int CAPACITY = 500;
+  private final Queue<String> knoxTokens = new ConcurrentLinkedQueue<>();
+
+  void saveKnoxToken(String knoxToken) {
+    cleanIfNecessary();
+    this.knoxTokens.offer(knoxToken);
```
Comment (pzampino) [inline gateway-performance-test/src/main/java/org/apache/knox/gateway/performance/test/knoxtoken/KnoxTokenCache.java:29]: Is it acceptable for a token to NOT be saved here? Just wondering why you're using _offer_ instead of _add_.

File: gateway-server/src/main/java/org/apache/knox/gateway/services/token/impl/DefaultTokenStateService.java [hunk 4]
```diff
@@ -154,17 +177,16 @@ public long getTokenExpiration(final String tokenId) throws UnknownTokenExceptio
...
 
+    long expiration = -1;
     synchronized (tokenExpirations) {
-      if (!tokenExpirations.containsKey(tokenId)) {
-        throw new UnknownTokenException(tokenId);
-      }
-      expiration = tokenExpirations.get(tokenId);
+      expiration = tokenExpirations.getOrDefault(tokenId, -1L);
+    }
+    if (expiration == -1) {
```
Comment (pzampino) [inline gateway-server/src/main/java/org/apache/knox/gateway/services/token/impl/DefaultTokenStateService.java:188]: This doesn't appear to be any different than the previous implementation. If this were a check for _expiration < 0_, I could see a difference. What is the reason for the change?

File: gateway-shell/src/main/java/org/apache/knox/gateway/shell/hdfs/Ls.java [hunk 5]
```diff
@@ -41,6 +42,11 @@ public Request dir( String dir ) {
       return this;
     }
 
+    public Request knoxToken(String knoxToken) {
```
Comment (pzampino) [inline gateway-shell/src/main/java/org/apache/knox/gateway/shell/hdfs/Ls.java:45]: Why not add Knox token support to KnoxSession?

File: gateway-performance-test/src/main/java/org/apache/knox/gateway/performance/test/knoxtoken/KnoxTokenCache.java (hunk 3 above)
Comment (smolnar82) [reply gateway-performance-test/src/main/java/org/apache/knox/gateway/performance/test/knoxtoken/KnoxTokenCache.java:29]: As discussed offline, it is acceptable: this cache works only as a storage to make sure there is an already acquired Knox DT to be renewed/used. It really does not matter if we miss some.

File: gateway-server/src/main/java/org/apache/knox/gateway/services/token/impl/DefaultTokenStateService.java (hunk 4 above)
Comment (smolnar82) [reply gateway-server/src/main/java/org/apache/knox/gateway/services/token/impl/DefaultTokenStateService.java:188]: There is a slight difference: we can save one iteration when checking `tokenExpirations`
`containsKey + get` vs. `getOrDefault`.
But you are right, there is no more difference between the two versions.

File: gateway-shell/src/main/java/org/apache/knox/gateway/shell/hdfs/Ls.java (hunk 5 above)
Comment (smolnar82) [reply gateway-shell/src/main/java/org/apache/knox/gateway/shell/hdfs/Ls.java:45]: IMO, the token handling should happen on request level. But you are right, this should not be limited to the `ls` command -> moving it to the abstract parent.

File: gateway-performance-test/src/main/java/org/apache/knox/gateway/performance/test/PerformanceTestRunner.java (hunk 2 above)
Comment (smolnar82) [reply gateway-performance-test/src/main/java/org/apache/knox/gateway/performance/test/PerformanceTestRunner.java]: Thanks for your comment; I made the requested change.

File: gateway-server/src/main/java/org/apache/knox/gateway/services/token/impl/DefaultTokenStateService.java (hunk 4 above)
Comment (smolnar82) [reply gateway-server/src/main/java/org/apache/knox/gateway/services/token/impl/DefaultTokenStateService.java:188]: I left it as-is based on my previous comment.

File: gateway-shell/src/main/java/org/apache/knox/gateway/shell/hdfs/Ls.java (hunk 5 above)
Comment (smolnar82) [reply gateway-shell/src/main/java/org/apache/knox/gateway/shell/hdfs/Ls.java]: Fixed.
